### PR TITLE
Service mirroring tests

### PR DESCRIPTION
**Pull request by zaharidichev**
Unit tests that exercise most of the code in cluster_watcher.go. Essentially the whole  cluster mirroring machinary can be tought of as a function that takes remote cluster state, local cluster state, and modification events and as a result it either modifies local cluster state or issues new events onto the queue. This is what these tests are trying to model. I think this covers a lot of the logic there. Any suggestions for other edge cases are welcome.

### CODE REVIEW

File: controller/cmd/service-mirror/cluster_watcher_test.go [hunk 1]
```diff
@@ -0,0 +1,1165 @@
...
+	}
+
+	if !reflect.DeepEqual(expected.Subsets, actual.Subsets) {
+		return fmt.Errorf("was expecting endpoints with subsets %v but got %v", expected.Subsets, actual.Subsets)
+	}
+
+	return nil
+}
+
+func getRemoteService(name, namespace, gtwName, gtwNs, resourceVersion string, ports []corev1.ServicePort) *corev1.Service {
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher_test.go]: since this isn't actually fetching anything from anywhere, should we call this `makeRemoteService` or even simply `remoteService`?

File: controller/cmd/service-mirror/cluster_watcher_test.go [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       namespace,
+			ResourceVersion: resourceVersion,
+			Annotations:     annotations,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: ports,
+		},
+		Status: corev1.ServiceStatus{},
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher_test.go]: Since this is the zero value, I think we can omit it.

File: controller/cmd/service-mirror/cluster_watcher_test.go (hunk 1 above)
Comment (adleong) [reply controller/cmd/service-mirror/cluster_watcher_test.go]: similar comment for other "get" functions in this file.

File: controller/cmd/service-mirror/cluster_watcher_test.go [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+func getEndpointsAsYaml(name, namespace, gtwName, gtwNs, gatewayIP string, ports []corev1.EndpointPort, t *testing.T) string {
+	ep := getEndpoints(name, namespace, gtwName, gtwNs, gatewayIP, ports)
+
+	bytes, err := yaml.Marshal(ep)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return fmt.Sprintf("%s", bytes)
+}
+
```
Comment (adleong) [inline controller/cmd/service-mirror/cluster_watcher_test.go]: I might consider moving all the helper methods (everything above this line, basically) into another file so that the test cases themselves are more front-and-center in this file.  TIOLI

File: controller/cmd/service-mirror/cluster_watcher_test.go [hunk 4]
```diff
@@ -0,0 +1,960 @@
...
+					Name:      "test-service-remote-to-delete",
+					Namespace: "test-namespace-to-delete",
+				},
+			},
+
+			localResources: []string{
+				mirroredServiceAsYaml("test-service-remote-to-delete", "test-namespace-to-delete", "", "", "", "", nil, t),
+				endpointsAsYaml("test-service-remote-to-delete", "test-namespace-to-delete", "", "", "", nil, t),
+			},
+		},
```
Comment (alpeb) [inline controller/cmd/service-mirror/cluster_watcher_test.go]: Given `expectedLocalServices` and `expectedLocalEndpoints` is empty in this test, the loops in `runTestCase()` won't run and there is no check performed verifying if that service and endpoint are really gone. 
One way of verifying could be checking if those expected slices are empty, and if so, loop against `localResources` and verify the local API can't retrieve any service/endpoint from that slice...

File: controller/cmd/service-mirror/cluster_watcher_test.go [hunk 5]
```diff
@@ -0,0 +1,960 @@
...
+			testDescription: "unregisters cluster and cleans up all mirrored resources",
+			events: []interface{}{
+				&ClusterUnregistered{},
+			},
+			localResources: []string{
+				mirroredServiceAsYaml("test-service-1-remote", "test-namespace", "", "", "", "", nil, t),
+				endpointsAsYaml("test-service-1-remote", "test-namespace", "", "", "", nil, t),
+				mirroredServiceAsYaml("test-service-2-remote", "test-namespace", "", "", "", "", nil, t),
+				endpointsAsYaml("test-service-2-remote", "test-namespace", "", "", "", nil, t),
+			},
```
Comment (alpeb) [inline controller/cmd/service-mirror/cluster_watcher_test.go]: Same here, it appears no check is performed.

File: controller/cmd/service-mirror/cluster_watcher_test.go (hunk 4 above)
Comment (zaharidichev) [reply controller/cmd/service-mirror/cluster_watcher_test.go]: Thanks for catching that @alpeb !
